### DISCUSSION TITLE
Calling a python script from another path

### DISCUSSION THREAD

**Discussion by nicoftrs** [the-anylogic-company/AnyLogic-Pypeline#44] [Q&A] [answered]
Hi,

Until now, I have been using Anylogic's pypeline features with Python files saved in the same directory as the Anylogic file which they are being called from. However, to tidy up my workspace, I'd like to have all my Python scripts saved in a separate directory and call them from there whenever I use Pypeline. This separate directory would be located at the same location as my Anylogic file. 

So in the example below, if I have a file called wind_power.py which has a function called wind_power, until now I call it like so, which works:

ArrayList output= pyCommunicator.runResults(ArrayList.class,String.format("wind_power.wind_power(%s,%s,%s,%s)", string1, string2, string3, string4));

Howere, I would like to save wind_power in a separate file called logic_files. I haven't figured out a way to call that file in that case. Would it be something like this?: 

 ArrayList output= pyCommunicator.runResults(ArrayList.class,String.format("from logic_files import wind_power.wind_power(%s,%s,%s,%s)", string1, string2, string3, string4));

Thank you in advance for any help provided. 

Kind regards,

Nicolas

**Comment by t-wolfeadam** [comment, marked as answer]
You import/call it the same way as in usual Python contexts: in terms of direct files, you can import modules (i.e., ".py" files) to get access to everything inside it (`import <module>`) or classes/functions within the module to be called directly (`from <module> import <function>`). I'm assuming in your current setup, you execute `import wind_power` somewhere.

As an example:
![image](https://github.com/t-wolfeadam/AnyLogic-Pypeline/assets/40703071/cf299309-181e-4f94-84ce-eac5464e74ea)

In your scenario, you say you want it in a separate directory but that "logic_files" is a file...? If you mean the "wind_power.py" file is in a directory named "logic_files", then that's also possible by creating a package. You'll need to add a blank "\_\_init\_\_.py" file within the folder for Python to detect it. And Python's import statements treat these like modules and you can use the "." in the "<module>" part to specify subpackages or modules within packages. 
So you'd do `from logic_files import wind_power` and then you'd call `wind_power.wind_power(...)` OR `from logic_files.wind_power import wind_power` and then you'd call `wind_power(...)`. 

In case it's not clear, as an example of all combinations, see the attached model/screenshot.

![image](https://github.com/t-wolfeadam/AnyLogic-Pypeline/assets/40703071/52aaa371-5e4c-49be-a3a9-a51cc6a3ac9f)
[Model288.zip](https://github.com/t-wolfeadam/AnyLogic-Pypeline/files/11523217/Model288.zip)

P.S. There is native string formatting available, so you can drop the "String.format" 😃  -- e.g., `ArrayList output =  pyCommunicator.runResults(ArrayList.class, "wind_power.wind_power(%s,%s,%s,%s)", string1, string2, string3, string4);`

**Reply by nicoftrs** [reply]
Brilliant, that worked, thank you very much for your quick reply!
